parser_library/src/workspaces/workspace.cpp

```diff
@@ -184,48 +253,58 @@ workspace_file_info workspace::parse_file(
     }
 
     // TODO: what about removing files??? what if depentands_ points to not existing file?
-    std::vector<processor_file_ptr> files_to_parse;
+    std::vector<processor_file_compoments*> files_to_parse;
 
     // TODO: apparently just opening a file without changing it triggers reparse
 
-    if (auto this_file = file_manager_.find_processor_file(file_location);
-        file_content_status == open_file_result::changed_content
-        || file_content_status == open_file_result::changed_lsp && !(this_file && this_file->has_lsp_info()))
+    if (processor_file_compoments* this_file = nullptr; file_content_status == open_file_result::changed_content
+        || file_content_status == open_file_result::changed_lsp
+            && !((this_file = find_processor_file_impl(file_location)) != nullptr
+                && this_file->m_processor_file->has_lsp_info()))
     {
         if (trigger_reparse(file_location))
             files_to_parse = collect_dependants(file_location);
 
-        if (files_to_parse.empty() && this_file)
-            files_to_parse.push_back(std::move(this_file));
+        if (files_to_parse.empty())
+        {
+            if (!this_file)
+                this_file = &add_processor_file_impl(file_location);
+            files_to_parse.push_back(this_file);
+        }
 
-        for (const auto& f : files_to_parse)
+        for (auto* component : files_to_parse)
         {
+            auto& f = component->m_processor_file;
             const auto& f_loc = f->get_location();
 
             auto alt_cfg = m_configuration.load_alternative_config_if_needed(f_loc);
             if (auto opened_it = opened_files_.find(f_loc); opened_it != opened_files_.end())
                 opened_it->second.alternative_config = std::move(alt_cfg);
 
-            if (!f->parse(*this, get_asm_options(f_loc), get_preprocessor_options(f_loc), &fm_vfm_))
+            workspace_parse_lib_provider ws_lib(*this, f_loc);
+            if (!f->parse(ws_lib, get_asm_options(f_loc), get_preprocessor_options(f_loc), &fm_vfm_))
                 continue;
 
-            ws_file_info = parse_successful(f);
+            ws_file_info = parse_successful(*component, ws_lib);
         }
 
         // second check after all dependants are there to close all files that used to be dependencies
-        for (const auto& f : files_to_parse)
-            filter_and_close_dependencies_(f->files_to_close(), f);
+        for (auto* component : files_to_parse)
+            filter_and_close_dependencies_(component->m_processor_file->files_to_close(), component->m_processor_file);
     }
 
     return ws_file_info;
 }
 
-workspace_file_info workspace::parse_successful(const processor_file_ptr& f)
+workspace_file_info workspace::parse_successful(processor_file_compoments& comp, workspace_parse_lib_provider& libs)
```

Comment:
```suggestion
workspace_file_info workspace::parse_successful(processor_file_compoments& comp, workspace_parse_lib_provider libs)
```

and using `std::move(...)` for `libs` when `parse_successful` is called on lines 199 and 288?